src/neuralnet/input_layer/store.cc

```diff
@@ -38,11 +39,21 @@ void StoreInputLayer::Setup(const LayerProto& conf,
   if (conf.partition_dim() == 0) {
     batchsize_ /= conf.num_partitions();
   }
+
+  vector<int> shape {batchsize_};
+  for (int s : conf.store_conf().shape())
+    shape.push_back(s);
+  data_.Reshape(shape);
+  aux_data_.resize(batchsize_);
+  buf_keys_.resize(batchsize_);
+  buf_vals_.resize(batchsize_);
+
+  // initialize prefetch buffer and start the thread
+  if (conf.store_conf().prefetching())
+    threads_.push_back(thread(&StoreInputLayer::fetch_data, this));
```

Comment:
For distributed training, we may have multiple StoreInputLayer.
All these layers are setup on every node to construct the whole neural net.
In that case, we will allocate many prefetch threads and buffers.However, they will not be used on every node.
E.g., input layer A and B will be setup on machine X and Y. But only X will run A and only Y will run B.
The solution is to start the prefetching inside ComputeFeature instead of Setup. X will call A's ComputeFeature, but not call B's ComputeFeature.